src/p2p/pubsub.rs

```diff
@@ -0,0 +1,352 @@
+    pub fn add_node_to_partial_view(&mut self, peer_id: PeerId) {
+        self.floodsub.add_node_to_partial_view(peer_id);
+    }
+
+    /// See [`Floodsub::remove_node_from_partial_view`]
+    pub fn remove_node_from_partial_view(&mut self, peer_id: &PeerId) {
+        self.floodsub.remove_node_from_partial_view(peer_id);
+    }
+}
+
+type PubsubNetworkBehaviourAction = NetworkBehaviourAction<
+    <<Pubsub as NetworkBehaviour>::ProtocolsHandler as ProtocolsHandler>::InEvent,
+    <Pubsub as NetworkBehaviour>::OutEvent,
+>;
+
+impl NetworkBehaviour for Pubsub {
+    type ProtocolsHandler = <Floodsub as NetworkBehaviour>::ProtocolsHandler;
+    type OutEvent = void::Void;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        self.floodsub.new_handler()
+    }
+
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        self.floodsub.addresses_of_peer(peer_id)
+    }
+
+    fn inject_connected(&mut self, peer_id: PeerId, cp: ConnectedPoint) {
+        self.floodsub.inject_connected(peer_id, cp)
+    }
+
+    fn inject_disconnected(&mut self, peer_id: &PeerId, cp: ConnectedPoint) {
+        self.floodsub.inject_disconnected(peer_id, cp)
+    }
+
+    fn inject_node_event(
+        &mut self,
+        peer_id: PeerId,
+        event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+    ) {
+        self.floodsub.inject_node_event(peer_id, event)
+    }
+
+    fn inject_addr_reach_failure(
+        &mut self,
+        peer_id: Option<&PeerId>,
+        addr: &Multiaddr,
+        error: &dyn std::error::Error,
+    ) {
+        self.floodsub
+            .inject_addr_reach_failure(peer_id, addr, error)
+    }
+
+    fn poll(
+        &mut self,
+        ctx: &mut Context,
+        poll: &mut impl PollParameters,
+    ) -> Poll<PubsubNetworkBehaviourAction> {
+        use futures::stream::StreamExt;
+        use std::collections::hash_map::Entry;
+
+        loop {
+            match self.unsubscriptions.1.poll_next_unpin(ctx) {
+                Poll::Ready(Some(dropped)) => {
+                    let topic = Topic::new(dropped);
+                    if self.streams.remove(&topic).is_some() {
+                        log::debug!("unsubscribing via drop from {:?}", topic.id());
+                        assert!(
+                            self.floodsub.unsubscribe(topic),
+                            "Failed to unsubscribe a dropped subscription"
+                        );
+                    } else {
+                        // unsubscribed already by `unsubscribe`
+                        // TODO: not sure if the unsubscribe functionality is needed as the
+                        // unsubscribe on drop seems to work
+                    }
+                }
+                Poll::Ready(None) => unreachable!("we own the sender"),
+                Poll::Pending => break,
+            }
+        }
+
+        match futures::ready!(self.floodsub.poll(ctx, poll)) {
```

Comment:
This needs to be in a loop, continue needs to be on where `Poll::Pending` is currently, otherwise the floodsub will have remaining pending events. Similar fix is needed to ipfs background task.

Reply:
Similar fix is needed at swarm (the `waker` can now be removed).

Reply:
I see, yes it makes sense that the daemon would wake the swarm task if needed